### DISCUSSION TITLE
Include path or directory

### DISCUSSION THREAD

**Discussion by jrouzierinverse** [kdl-org/kdl#406] [Ideas]
Is it possible to add support for include files?

A directory
`@include "/config.d/*"`

A single file
`@include "/extra.kdl"`

**Comment by IceDragon200** [comment]
```elixir
iex(1)> Kuddle.decode("""
...(1)> @include "/config.d/*"
...(1)> """)
{:ok,
 [
   %Kuddle.Node{
     name: "@include",
     annotations: [],
     attributes: [
       %Kuddle.Value{
         value: "/config.d/*",
         annotations: [],
         type: :string,
         format: :plain
       }
     ],
     children: nil
   }
 ], []}
```

Yes? You can technically do that as evidenced by me parsing it.
As a functionality of KDL, No, KDL does not provide "functions", it's up to whatever application is consuming the KDL document to understand the functional nodes (e.g. `@include`) and then process them as seen.

From a security standpoint: that is a dangerous feature to include willy-nilly.

EDIT:
Mister include, if you see this message and wonder why you were pinged, I apologize for not wrapping the `@include` initially in a code fence.

**Comment by zkat** [comment]
To expand on this: KDL is not necessarily consumed from files, and different applications may want different semantics for their `@include` equivalent. I think it's reasonable, and not onerous, to leave this kind of work to the application layer in order to keep the spec itself simple (including not inherently dealing with security and filesystem issues), and to give application developers the flexibility they need to define both syntax and semantics.
